File: Config1/UNOAUV1/FPO/TABTS.cc

#include "globals.h" 
#include "TABTS.h" 

    int SMC = 0;		 //Счетчик периода мерцания
    bool PMC = false;		 //Признак мерцания
    int SMG = 0;		 //Счетчик периода мигания
    bool PMG  =false;	 	 //Признак мигания

    bool _PI12AB01PTS = false;
    bool _PI12AB05PTS = false;
    bool _PI12AB09PTS = false;
 
    bool _PI12AB01PRMC = false;
    bool _PI12AB05PRMC = false;
    bool _PI12AB09PRMC = false;
 
    bool _PI12AB01PRMG = false;
    bool _PI12AB05PRMG = false;
    bool _PI12AB09PRMG = false;
 
    bool _PI12AB01PRRP = false;
    bool _PI12AB05PRRP = false;
    bool _PI12AB09PRRP = false;
 
    int TT = 100;     //Период задачи	

  void  _TABTSS()
 {

    int TMC = 500/TT;	 //Период мерцания
    int TMG = 2000/TT;	 //Период мигания

  int n = 3;        //Кол-во  сообщений ТС 
  bool* TS[3] = {    //сообщение ТС
   &_PI12AB01TS,
   &_PI12AB05TS,
   &_PI12AB09TS
  };
  bool* DTS[3] = {    //сигнал на табло
   &_PI12AB01DTS,
   &_PI12AB05DTS,
   &_PI12AB09DTS
  };
   bool KNMC = _PI12AG03KNMC;  //кнопка мерцания 
   bool KNMG = _PI12AG05KNMG;  //кнопка мигания 
  bool* PTS[3] = {    //Признак i-го сообщения ТС
   &_PI12AB01PTS,
   &_PI12AB05PTS,
   &_PI12AB09PTS
  };
  bool* PRMC[3] = {    //Признак мерцания i-го сообщения ТС
   &_PI12AB01PTS,
   &_PI12AB05PTS,
   &_PI12AB09PTS
  };
  bool* PRMG[3] = {    //Признак мигания i-го сообщения ТС
   &_PI12AB01PTS,
   &_PI12AB05PTS,
   &_PI12AB09PTS
  };
  bool* PRR[3] = {    //Признак ровного свечения i-го сообщения ТС
   &_PI12AB01PTS,
   &_PI12AB05PTS,
   &_PI12AB09PTS
  };

int i;
 
  if (SMC++ == TMC)
  {
    SMC = 0;
    PMC = !PMC;
  }

  if (SMG++ == TMG)
  {
    SMG = 0;
    PMG = !PMG;
   }

 for (i = 0; i < n; ++i) 
  {
     if (KNMC || *PRR[i]) 
       { if (*TS[i])
            { *DTS[i]=*PRR[i]=*PTS[i]=true; }
	
          else 
		{
		 if (*PTS[i])
			{ *PRMG[i]=true; *PRR[i]=*PRMC[i]=false; *DTS[i]=PMG; }
 		  else  *PTS[i]=*DTS[i]=*PRMG[i]=*PRMC[i]=false;
		}
       }
     else
     { //1
      if (!*PRMG[i] && (*PRMC[i] || *TS[i]))
		{ *PTS[i]=*PRMC[i]=true; *PRMG[i]=*PRR[i]=false; *DTS[i]=PMC; }
	else
        { //2
         if ((!*PRMG[i] && !*PRMC[i] && !*TS[i]) || (*PRMG[i] && KNMG))
		*PTS[i]=*DTS[i]=*PRMG[i]=*PRMC[i]=false;
	  else
	    { //3
             if (*TS[i])
	        { *PTS[i]=*PRMC[i]=true; *PRMG[i]=*PRR[i]=false; *DTS[i]=PMC; }
               else
                { *PRMG[i]=true; *PRR[i]=*PRMC[i]=false; *DTS[i]=PMG; }
            } //3
	}//2

     }//1

  }
}

File: Scripts/TABTS.php
<?
//Параметры: путь к разработке, контроллер, период задачи
$params = "lbd,takt"; 
include "sintar2007.php";
?>

#include "globals.h" 
#include "TABTS.h" 

    int SMC = 0;		 //Счетчик периода мерцания
    bool PMC = false;		 //Признак мерцания
    int SMG = 0;		 //Счетчик периода мигания
    bool PMG  =false;	 	 //Признак мигания

<?
global $signals;
$sigdb = ibase_connect($signals);
$cont = substr(basename($lbd,'.ib'),3); //Имя контроллера
//Выбираем сигналы DO и PFD, принадлежащие заданной ЛБД
$qry = "select SC.SigName from SigCont SC, Controller C
  where C.Name='{$cont}' and SC.NodeID=C.NodeID 
  and (SC.TypeName='DI' or SC.TypeName='DO' or SC.TypeName='PFD')
  order by SC.SigName";
$sig_q = ibase_query($sigdb, $qry);

//для подсветки ТАБЛО ТС  4 сигнала  из БД:
$ts = array();		//сообщение сигнализации
$dts = array();		//сигнал на табло
$knmc = '';		//кнопка мерцания
$knmg = '';		//кнопка мигания
// 4 ПРИЗНАКА СООБЩЕНИЯ ТС
$pts  = array();
$prmc = array();
$prmg = array();
$prr  = array();


while ($sig = ibase_fetch_object($sig_q))
{
//echo "{$sig->SIGNAME}\n";
  
  $tsname = substr($sig->SIGNAME,0,8);

  $proto = substr($sig->SIGNAME,8);   // $proto - возвращает строку после ИМ

  if ($proto == 'TS')
   {   
   $ts[]   = $sig->SIGNAME;
   $pts[]  = str_pad($tsname, 11, 'PTS');	
   $prmc[] = str_pad($tsname, 12, 'PRMC');
   $prmg[] = str_pad($tsname, 12, 'PRMG');
   $prr[]  = str_pad($tsname, 12, 'PRR');
   }	
  elseif ($proto == 'DTS')
    $dts[] = $sig->SIGNAME;	

  elseif ($proto == 'KNMC')
    $knmc = $sig->SIGNAME;	
  elseif ($proto == 'KNMG')
    $knmg = $sig->SIGNAME;	
 	
}
$n = count($ts);

out2sig_arr($pts);
out2sig_arr($prmc);
out2sig_arr($prmg);
out2sig_arr($prr);

echo "    int TT = {$takt};     //Период задачи	\n";

?>

  void  _TABTSS()
 {

    int TMC = 500/TT;	 //Период мерцания
    int TMG = 2000/TT;	 //Период мигания

<?

echo "  int n = {$n};        //Кол-во  сообщений ТС \n";

out1sig_arr($ts, 'TS',$n, 'сообщение ТС');
out1sig_arr($dts, 'DTS',$n, 'сигнал на табло');

echo "   bool KNMC = _{$knmc};  //кнопка мерцания \n";
echo "   bool KNMG = _{$knmg};  //кнопка мигания \n";

out1sig_arr($pts, 'PTS',  $n, 'Признак i-го сообщения ТС');
out1sig_arr($pts, 'PRMC', $n, 'Признак мерцания i-го сообщения ТС');
out1sig_arr($pts, 'PRMG', $n, 'Признак мигания i-го сообщения ТС');
out1sig_arr($pts, 'PRR',  $n, 'Признак ровного свечения i-го сообщения ТС');

?>

int i;
 
  if (SMC++ == TMC)
  {
    SMC = 0;
    PMC = !PMC;
  }

  if (SMG++ == TMG)
  {
    SMG = 0;
    PMG = !PMG;
   }

 for (i = 0; i < n; ++i) 
  {
     if (KNMC || *PRR[i]) 
       { if (*TS[i])
            { *DTS[i]=*PRR[i]=*PTS[i]=true; }
	
          else 
		{
		 if (*PTS[i])
			{ *PRMG[i]=true; *PRR[i]=*PRMC[i]=false; *DTS[i]=PMG; }
 		  else  *PTS[i]=*DTS[i]=*PRMG[i]=*PRMC[i]=false;
		}
       }
     else
     { //1
      if (!*PRMG[i] && (*PRMC[i] || *TS[i]))
		{ *PTS[i]=*PRMC[i]=true; *PRMG[i]=*PRR[i]=false; *DTS[i]=PMC; }
	else
        { //2
         if ((!*PRMG[i] && !*PRMC[i] && !*TS[i]) || (*PRMG[i] && KNMG))
		*PTS[i]=*DTS[i]=*PRMG[i]=*PRMC[i]=false;
	  else
	    { //3
             if (*TS[i])
	        { *PTS[i]=*PRMC[i]=true; *PRMG[i]=*PRR[i]=false; *DTS[i]=PMC; }
               else
                { *PRMG[i]=true; *PRR[i]=*PRMC[i]=false; *DTS[i]=PMG; }
            } //3
	}//2

     }//1

  }
}


